### PR TITLE
[In_App_Purchase] Improve testability

### PR DESCRIPTION
**Pull request by BugsBunnyBR**
## Description
This PR creates a factory for BillingClient, with that was possible to remove a secondary constructor that was used just for test.

With the introduction of the factory there was the need to update some tests as it was testing states that would never be reached in the normal plugin flow (due the usage of the constructor for testing). Basically, the test cases depended on the state of BillingClient that would be null if no previous setup was executed.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [X] All existing and new tests are passing.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [X] I read and followed the [Flutter Style Guide].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I signed the [CLA].
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/in_app_purchase/android/src/main/java/io/flutter/plugins/inapppurchase/BillingClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package io.flutter.plugins.inapppurchase;
+
+import android.content.Context;
+import com.android.billingclient.api.BillingClient;
+import io.flutter.plugin.common.MethodChannel;
+
+public class BillingClientFactory {
```
[the remark this answers is not in the document]
Comment (BugsBunnyBR) [reply packages/in_app_purchase/android/src/main/java/io/flutter/plugins/inapppurchase/BillingClientFactory.java]: I think all your suggestions are fair.
Mockito v2 can mock final classes! \o/
https://stackoverflow.com/a/40018295/2742962
but it comes with some overhead /o\
As we control the abstraction, let's go with the interface. :-D

**Comment by BugsBunnyBR** [comment]
@mklim 
I just ran `pub global run flutter_plugin_tools format` and now the CI is complaining. Would you help me? I am using flutter master channel here.

**Comment by cyanglaz** [comment]
It seems to complain about a recent change I made in the connectivity. 
https://github.com/flutter/plugins/commit/90aaca84a697fb1cf714e8570cdde0c38c2fda11 didn't complain about the formatting tho. I will keep an eye out.

**Comment by cyanglaz** [comment]
@BugsBunnyBR I think the commit passed the formatting check. Could you try again after merging master, and also make sure your flutter is on master?

**Comment by BugsBunnyBR** [comment]
> @BugsBunnyBR I think the commit passed the formatting check. Could you try again after merging master, and also make sure your flutter is on master?

This branch is even with the master branch, I also ran `pub global run flutter_plugin_tools format` from flutter channel master.

**Comment by mklim** [comment]
@BugsBunnyBR @cyanglaz it looks for some reason the formatting check is just failing now. We have logic to only run checks against "changed" plugins which must have failed here. Or it's possible that the formatting rules changed since the PRs were merged in. I'll submit a fix to unblock this for now.

**Comment by mklim** [comment]
@BugsBunnyBR it looks like the formatter is complaining about a change it applied through you to this branch, a07f6e26e09ca575fb1fbeac7e7cd7f12e4be670. I bet the problem is the same that I had, running `pub global run flutter_plugin_tools format` with a different version of `clang-format` than CI on your PATH. I fixed this locally by installing `clang-format-7` and passing it to the tool, so `pub global run flutter_plugin_tools format --clang-format=clang-format-7`.

**Comment by BugsBunnyBR** [comment]
@mklim Thank you! Let's see if the CI likes the new commit.

**Pull request merged**